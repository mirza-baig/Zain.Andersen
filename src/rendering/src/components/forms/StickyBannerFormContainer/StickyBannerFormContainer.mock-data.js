const defaultData = {
  rendering: {
    componentName: 'StickyBannerFormContainer',
    dataSource: 'sampledatasource',
  },
  fields: {
    ctaText: {
      value: 'Start Here',
    },
    componentSpacing: {
      id: 'a1c45922-1caf-4df8-9f96-9d2033e159cf',
      url: '/sitecore/content/andersencorporation/enterprise-global/enums/component-spacing/none',
      name: 'None',
      displayName: 'None',
      fields: {
        Value: {
          value: 'none',
        },
      },
      templateId: 'd2923fee-da4e-49be-830c-e27764dfa269',
      templateName: 'Enum',
    },
    sectionId: {
      value: '',
    },
    headlineLevel: {
      id: '0f556f3a-bbad-4aa5-952d-b79003b39cd6',
      url: '/sitecore/content/andersencorporation/enterprise-global/enums/headline-levels/heading-2',
      name: 'Heading 2',
      displayName: 'Heading 2',
      fields: {
        Value: {
          value: 'h2',
        },
      },
      templateId: 'd2923fee-da4e-49be-830c-e27764dfa269',
      templateName: 'Enum',
    },
    headlineText: {
      value: 'Take the first step',
    },
    eventName: {
      value: '',
    },
    eventType: {
      value: '',
    },
    eventZone: {
      value: '',
    },
  },
};

export default defaultData;
