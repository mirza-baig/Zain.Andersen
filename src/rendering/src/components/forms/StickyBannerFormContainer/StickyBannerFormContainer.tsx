// Global
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { useTheme } from 'lib/context/ThemeContext';
import { Placeholder, withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
// Components
import { Component } from 'src/helpers/Component';
import classNames from 'classnames';
import { useEffect, useState } from 'react';
import { Headline } from 'src/helpers/Headline';
import { getEnum } from 'lib/utils';
import { ButtonPrimaryClasses } from 'src/helpers/Button/buttons/btn--primary';
import { SvgIcon } from 'src/helpers/SvgIcon';
import ModalWrapper from 'src/helpers/ModalWrapper/ModalWrapper';
import { getBreakpoint, useCurrentScreenType } from 'lib/utils/get-screen-type';
import { getCookie } from 'cookies-next';
import { Subheadline } from 'src/helpers/Subheadline';
import { useScrollDirection } from 'lib/utils/use-scroll-direction';

export type StickyBannerFormContainerProps =
  Feature.EnterpriseWeb.Enterprise.Components.General.StickyBanner.StickyBanner;
const StickyBannerFormContainer = (props: StickyBannerFormContainerProps) => {
  const { fields } = props;
  const { themeName } = useTheme();
  const { currentScreenWidth } = useCurrentScreenType();
  const scrollDirection = useScrollDirection(0);

  // Access the cookie value
  const JOB_INQUIRY_SUBMITTED_COOKIE_NAME = 'JobInquirySubmitted';

  const isMobile = currentScreenWidth < getBreakpoint('md');

  const [isFormSticky, setIsFormSticky] = useState(false);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [isCookieAvailable, setIsCookieAvailable] = useState(false);

  useEffect(() => {
    setIsCookieAvailable(() => (getCookie(JOB_INQUIRY_SUBMITTED_COOKIE_NAME) ? true : false));

    const handleScroll = () => {
      if (
        window.scrollY > 3 ||
        isMobile ||
        scrollDirection === 'DOWN' ||
        scrollDirection === 'UP'
      ) {
        setIsFormSticky(true);
      } else if (window.scrollY <= 5) {
        setIsFormSticky(false);
      } else {
        setIsFormSticky(false);
      }
    };

    window.addEventListener('scroll', handleScroll);

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [isFormSticky, isMobile, isModalOpen]);

  if (!fields || isCookieAvailable) {
    return <></>;
  }

  const headlineLevel = getEnum<string>(props?.fields.headlineLevel);

  const handleCloseModal = () => {
    setIsModalOpen(!isModalOpen);
  };

  return (
    <Component
      dataComponent="forms/stickybannerformcontainer"
      variant="full"
      sectionWrapperClasses={classNames(
        isFormSticky ? 'ml:top-0 transition-all ease-in-out duration-500' : 'ml:top-[64px]',
        isModalOpen && '!fixed !top-0 transition-none',
        scrollDirection === 'UP' && 'ml:top-[80px] transition-all ease-in-out duration-300',
        'fixed ml:sticky bottom-0 left-0 right-0 block z-[1000] h-[64px] '
      )}
      id="stickybannerformcontainer"
      {...props}
    >
      <div
        className={classNames(
          !isModalOpen && 'ml:hover:h-[72px]',
          'col-span-12 mx-[calc(50%-50vw)] flex min-h-[64px] cursor-pointer items-center bg-primary'
        )}
        onClick={handleCloseModal}
      >
        <div className="mx-auto w-full max-w-screen-lg">
          <div className="flex w-full flex-row !items-center justify-center">
            <Headline
              defaultTag={headlineLevel}
              {...props}
              classes="text-xs font-bold font-serif mr-m"
            />
            <button
              onClick={handleCloseModal}
              className={classNames(
                ButtonPrimaryClasses(themeName).btnClass,
                '!h-[48px] !bg-white !text-black hover:!bg-white hover:!text-black'
              )}
            >
              <Subheadline
                classes="!text-button !font-serif font-bold"
                fields={{
                  subheadlineText: {
                    value: fields.ctaText.value,
                  },
                }}
              />
              <SvgIcon
                icon="arrow"
                className={classNames(
                  ButtonPrimaryClasses(themeName).iconClass,
                  '!text-primary hover:!text-primary'
                )}
              />
            </button>
          </div>
        </div>
      </div>

      {/* Modal */}
      {/* rename existing customWrapperClass -> customOverlayclass */}
      {isModalOpen && (
        <ModalWrapper
          customOverlayclass={'mt-[64px]'}
          customContentWrapperclass={'!bg-light-gray'}
          isModalOpen={isModalOpen}
          handleClose={handleCloseModal}
          size="extra-large"
          closeIconClasses="mt-xs mb-xxxs mr-s !bg-light-gray text-black"
        >
          <div className="col-span-12 mx-[calc(50%-50vw)] flex min-h-[64px] cursor-pointer items-center bg-light-gray">
            <div className="mx-auto w-full max-w-screen-lg">
              <div className="flex h-full w-full flex-row !items-center justify-center">
                <div className="z-[2] -mb-s flex flex-col items-center justify-evenly gap-y-s overflow-x-scroll bg-light-gray px-m py-s !text-black">
                  <Placeholder
                    isGenericModal={isModalOpen}
                    name="components"
                    rendering={props.rendering}
                    render={(components) => {
                      return components.map((component, index) => (
                        <div key={index}>{component}</div>
                      ));
                    }}
                  />
                </div>
              </div>
            </div>
          </div>
        </ModalWrapper>
      )}
    </Component>
  );
};

export default withDatasourceCheck()<StickyBannerFormContainerProps>(StickyBannerFormContainer);
