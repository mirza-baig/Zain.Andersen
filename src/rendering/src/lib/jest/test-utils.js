// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
