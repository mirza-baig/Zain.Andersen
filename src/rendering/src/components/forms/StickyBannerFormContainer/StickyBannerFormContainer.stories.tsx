// Global
import { Story, Meta } from '@storybook/react';
// Local
import StickyBannerFormContainer, {
  StickyBannerFormContainerProps,
} from './StickyBannerFormContainer';
import defaultData from './StickyBannerFormContainer.mock-data';

export default {
  title: 'Forms/StickyBannerFormContainer',
  component: StickyBannerFormContainer,
} as Meta;
const Template: Story<StickyBannerFormContainerProps> = (props) => (
  <StickyBannerFormContainer {...props} />
);
export const Default = Template.bind({});
Default.args = defaultData;
