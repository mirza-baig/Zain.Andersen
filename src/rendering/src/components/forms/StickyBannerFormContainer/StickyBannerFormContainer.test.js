// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import StickyBannerFormContainer from './StickyBannerFormContainer';
import defaultData from './StickyBannerFormContainer.mock-data';

it('renders correctly', async () => {
  const component = snapshot(StickyBannerFormContainer, { componentProps: defaultData });
  await hasDataComponent(component, 'forms/stickybannerformcontainer');
});
