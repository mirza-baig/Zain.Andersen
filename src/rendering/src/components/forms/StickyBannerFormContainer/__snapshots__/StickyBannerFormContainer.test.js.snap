// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  class="ml:top-[64px] ml:top-[80px] transition-all ease-in-out duration-300 fixed ml:sticky bottom-0 left-0 right-0 block z-[1000] h-[64px] "
  data-component="forms/stickybannerformcontainer"
  id="id1695242607"
>
  <div
    class="my-0 w-full theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s "
    >
      <div
        class="ml:hover:h-[72px] col-span-12 mx-[calc(50%-50vw)] flex min-h-[64px] cursor-pointer items-center bg-primary"
      >
        <div
          class="mx-auto w-full max-w-screen-lg"
        >
          <div
            class="flex w-full flex-row !items-center justify-center"
          >
            <div
              class="items-top font-sans [&_a:hover]:underline text-xs font-bold font-serif mr-m"
            >
              <h2>
                Take the first step
              </h2>
            </div>
            <button
              class="flex w-fit items-center whitespace-normal rounded-lg border-4 border-theme-btn-border bg-theme-btn-bg px-m py-[9px] font-sans text-button font-heavy text-theme-btn-text hover:border-theme-btn-border-hover hover:bg-theme-btn-bg-hover hover:text-theme-btn-text-hover disabled:border-gray disabled:text-gray disabled:hover:bg-theme-btn-bg disabled:cursor-not-allowed !h-[48px] !bg-white !text-black hover:!bg-white hover:!text-black"
            >
              <div
                class="font-sans !text-button !font-serif font-bold"
              >
                <h3>
                  Start Here
                </h3>
              </div>
              <span
                class="ml-xxs !text-primary hover:!text-primary"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
